src/sema/yul/expression.rs

```diff
@@ -115,10 +117,52 @@ fn resolve_bool_literal(
 
 fn resolve_number_literal(
     loc: &pt::Loc,
-    value: &BigInt,
+    base: &str,
+    exp: &str,
     ty: &Option<pt::Identifier>,
     ns: &mut Namespace,
 ) -> Result<YulExpression, ()> {
+    let base = BigInt::from_str(base).unwrap();
+
+    let value = if exp.is_empty() {
+        base
+    } else {
+        let b10 = BigInt::from_str("10").unwrap();
+
+        if let Some(exp) = exp.strip_prefix('-') {
+            if let Ok(exp) = u8::from_str(exp) {
+                let exp = b10.pow(exp);
+                let res = BigRational::new(base, exp);
+
+                if res.is_integer() {
+                    res.to_integer()
+                } else {
+                    ns.diagnostics.push(Diagnostic::error(
+                        *loc,
+                        "rational numbers not permitted".to_string(),
+                    ));
+                    return Err(());
+                }
+            } else {
+                ns.diagnostics.push(Diagnostic::error(
+                    *loc,
+                    format!("exponent '{}' too large", exp),
+                ));
+                return Err(());
+            }
+        } else if let Ok(exp) = u8::from_str(exp) {
+            let exp = b10.pow(exp);
```

Comment:
Same as above.

Reply:
True. I don't know what to call it, so I've removed the variable `let exp = `